File: cpp/src/offtarget.cpp
#include "offtarget.hpp"

#include <stdexcept>
#include <string>

namespace crispritz {

// =========================================================================
// Free functions
// =========================================================================

Strand strand_from_char(char c) {
  if (c == '+')
    return Strand::Forward;
  if (c == '-')
    return Strand::Reverse;
  throw std::invalid_argument(
      std::string("strand_from_char: expected '+' or '-', got '") + c + '\'');
}

// =========================================================================
// Constructor
// =========================================================================

OffTarget::OffTarget(std::string chrom, int32_t pos, Strand strand,
                     std::string grna, std::string target, int mismatches,
                     int bulge_dna, int bulge_rna)
    : chrom_(std::move(chrom)), pos_(pos), strand_(strand),
      grna_(std::move(grna)), target_(std::move(target)),
      mismatches_(mismatches), bulge_dna_(bulge_dna), bulge_rna_(bulge_rna) {
  // Validate after moving so error messages can still reference the values.
  if (chrom_.empty())
    throw std::invalid_argument("OffTarget: chrom must not be empty");

  if (pos_ <= 0)
    throw std::invalid_argument(
        "OffTarget: pos must be > 0 (1-based coordinate), got " +
        std::to_string(pos));

  if (grna_.empty())
    throw std::invalid_argument("OffTarget: grna must not be empty");

  if (target_.empty())
    throw std::invalid_argument("OffTarget: target must not be empty");

  if (mismatches_ < 0)
    throw std::invalid_argument("OffTarget: mismatches must be >= 0, got " +
                                std::to_string(mismatches));

  if (bulge_dna_ < 0)
    throw std::invalid_argument("OffTarget: bulge_dna must be >= 0, got " +
                                std::to_string(bulge_dna));

  if (bulge_rna_ < 0)
    throw std::invalid_argument("OffTarget: bulge_rna must be >= 0, got " +
                                std::to_string(bulge_rna));
}

// =========================================================================
// Derived properties
// =========================================================================

std::string OffTarget::bulge_type() const {
  // Derived from bulge_dna_ / bulge_rna_ — no separate stored field
  // to keep the class invariant simple (the three fields always agree).
  if (bulge_dna_ == 0 && bulge_rna_ == 0)
    return "X";
  if (bulge_dna_ > 0 && bulge_rna_ == 0)
    return "DNA";
  if (bulge_dna_ == 0 && bulge_rna_ > 0)
    return "RNA";
  return "DNA,RNA";
}

// =========================================================================
// Comparison operators
// =========================================================================

bool OffTarget::operator==(const OffTarget &other) const noexcept {
  // Locus identity only — guide sequence and edit counts are excluded.
  // See class-level documentation for the rationale.
  return chrom_ == other.chrom_ && pos_ == other.pos_ &&
         strand_ == other.strand_ && target_ == other.target_;
}

bool OffTarget::operator!=(const OffTarget &other) const noexcept {
  return !(*this == other);
}

bool OffTarget::operator<(const OffTarget &other) const noexcept {
  // Priority: chrom → pos → strand → total edit distance.
  //
  // std::tie cannot be used here because total_edit_distance() is a
  // computed value (not a stored l-value), so a cascaded if-chain is
  // clearer and avoids temporary storage.
  if (chrom_ != other.chrom_)
    return chrom_ < other.chrom_;

  if (pos_ != other.pos_)
    return pos_ < other.pos_;

  // Strand: '+' (ASCII 43) < '-' (ASCII 45).
  if (strand_ != other.strand_)
    return to_char(strand_) < to_char(other.strand_);

  return total_edit_distance() < other.total_edit_distance();
}

// =========================================================================
// Serialization
// =========================================================================

std::string OffTarget::to_tsv_row(char sep) const {
  // Build via string concatenation with a pre-reserved capacity.
  // Avoids the construction overhead and locale state of std::ostringstream
  // for what is a simple, fixed-schema row.
  std::string row;
  row.reserve(chrom_.size() + grna_.size() + target_.size() + 64u);

  row += chrom_;
  row += sep;
  row += std::to_string(pos_);
  row += sep;
  row += to_char(strand_);
  row += sep;
  row += grna_;
  row += sep;
  row += target_;
  row += sep;
  row += std::to_string(mismatches_);
  row += sep;
  row += std::to_string(bulge_dna_);
  row += sep;
  row += std::to_string(bulge_rna_);
  row += sep;
  row += bulge_type();

  return row;
}

std::vector<std::string> OffTarget::tsv_header() {
  // Column order mirrors to_tsv_row().  CFD score is deliberately absent;
  // the Python scoring layer appends it after computing it over the rows
  // returned from the C++ search.
  return {"chrom",      "pos",       "strand",    "grna",      "target",
          "mismatches", "bulge_dna", "bulge_rna", "bulge_type"};
}

// =========================================================================
// Debug / diagnostic
// =========================================================================

std::string OffTarget::locus_string() const {
  // E.g. "chr1:123456(+)"
  return chrom_ + ':' + std::to_string(pos_) + '(' + to_char(strand_) + ')';
}

} // namespace crispritz

File: src/crispritz_plus/offtarget.py
"""Representation of a single CRISPR off-target site.

This module defines :class:`OffTarget`, the canonical in-memory representation
of one off-target candidate produced by the TST search.  Every instance maps
directly to one row of the final TSV report and carries all fields needed for
downstream scoring (CFD, CRISTA, ...).
"""

from __future__ import annotations

from typing import List, Optional, Tuple


# ==============================================================================
# Constants
# ==============================================================================

#: Sentinel used for scores that have not yet been computed.
_SCORE_NA: str = "NA"

#: TSV column order – used by :py:meth:`OffTarget.to_tsv_row`.
TSV_HEADER: List[str] = [
    "chrom",
    "pos",
    "strand",
    "grna",
    "spacer",
    "mismatches",
    "bulge_dna",
    "bulge_rna",
    "bulge_type",
    "cfd_score",
]


# ==============================================================================
# OffTarget class
# ==============================================================================


class OffTarget:
    """Represent a single CRISPR off-target site and its associated scores.

    Each instance corresponds to one candidate off-target locus produced by the
    TST near-search.  The object is intentionally lightweight: it stores raw
    alignment data and lazily-populated scores, and knows how to serialise
    itself to a TSV row for the final report.

    Attributes
    ----------
    _chrom : str
        Chromosome (or contig) name of the off-target site.
    _pos : int
        1-based genomic start position.
    _strand : str
        Strand on which the site was found: ``'+'`` or ``'-'``.
    _grna : str
        Aligned guide RNA sequence including PAM placeholder ``N``.
        Mismatch positions are represented in upper-case; gap characters
        (``'-'``) mark bulge positions.
    _spacer : str
        Aligned genomic target (spacer) sequence including PAM bases.
        Mismatch bases are lower-case (convention from the C++ search).
    _mm : int
        Number of substitution mismatches between guide and target.
    _bulge_type : str
        Bulge classification: ``'X'`` (none), ``'DNA'``, ``'RNA'``, or
        ``'DNA,RNA'``.
    _bdna : int
        Number of DNA-bulge bases (gap in guide / extra base in target).
    _brna : int
        Number of RNA-bulge bases (gap in target / extra base in guide).
    _cfd_score : str
        Cutting Frequency Determination score, or ``'NA'`` when not yet
        computed.  Stored as a string to allow ``'NA'`` as a sentinel
        without requiring ``Optional[float]`` throughout the pipeline.
    _debug : bool
        When ``True``, exceptions propagate with full tracebacks instead of
        being swallowed into user-friendly error messages.

    Examples
    --------
    >>> ot = OffTarget(
    ...     chrom="chr1",
    ...     pos=123456,
    ...     strand="+",
    ...     grna="ACGTACGTACGTACGTACGTNGG",
    ...     spacer="ACGTACGTACGTACGTACGTaGG",
    ...     mm=1,
    ...     bulge_type="X",
    ...     bdna=0,
    ...     brna=0,
    ... )
    >>> ot.cfd_score = "0.85"
    >>> print(ot.to_tsv_row())
    chr1\t123456\t+\tACGTACGTACGTACGTACGTNGG\tACGTACGTACGTACGTACGTaGG\t1\tX\t0\t0\t0.85
    """

    # ==========================================================================
    # Construction
    # ==========================================================================

    def __init__(
        self,
        chrom: str,
        pos: int,
        strand: str,
        grna: str,
        spacer: str,
        mm: int,
        bulge_type: str,
        bdna: int,
        brna: int,
        cfd_score: str = _SCORE_NA,
        debug: bool = False,
    ) -> None:
        """Initialise an :class:`OffTarget` instance.

        Parameters
        ----------
        chrom:
            Chromosome name (e.g. ``'chr1'``).
        pos:
            1-based genomic start position (must be > 0).
        strand:
            Strand indicator; must be ``'+'`` or ``'-'``.
        grna:
            Aligned guide RNA sequence with PAM (``N``-placeholder) appended
            or prepended depending on the PAM orientation.
        spacer:
            Aligned genomic target sequence with PAM bases.  Lower-case
            characters indicate mismatch positions; ``'-'`` indicates bulge
            gaps.
        mm:
            Number of substitution mismatches (>= 0).
        bulge_type:
            One of ``'X'``, ``'DNA'``, ``'RNA'``, ``'DNA,RNA'``.
        bdna:
            Number of DNA-bulge bases (>= 0).
        brna:
            Number of RNA-bulge bases (>= 0).
        cfd_score:
            Pre-computed CFD score string, or ``'NA'`` (default).
        debug:
            Enable debug mode for verbose error propagation.

        Raises
        ------
        ValueError
            If any argument fails basic sanity checks (invalid strand, negative
            counts, unknown bulge type, …).
        """
        self._debug = debug
        self._chrom = chrom
        self._pos = pos
        self._strand = strand
        self._grna = grna
        self._spacer = spacer
        self._mm = mm
        self._bulge_type = bulge_type
        self._bdna = bdna
        self._brna = brna
        self._cfd_score: str = cfd_score

    # ------------------------------------------------------------------
    # Representation
    # ------------------------------------------------------------------

    def __repr__(self) -> str:  # noqa: D105
        return (
            f"OffTarget("
            f"chrom={self._chrom!r}, "
            f"pos={self._pos}, "
            f"strand={self._strand!r}, "
            f"mm={self._mm}, "
            f"bulge_type={self._bulge_type!r}, "
            f"bdna={self._bdna}, "
            f"brna={self._brna}, "
            f"cfd={self._cfd_score!r}"
            f")"
        )

    def __str__(self) -> str:  # noqa: D105
        return (
            f"{self._chrom}:{self._pos}{self._strand}  "
            f"gRNA={self._grna}  spacer={self._spacer}  "
            f"mm={self._mm}  bulge={self._bulge_type}({self._bdna}/{self._brna})  "
            f"CFD={self._cfd_score}"
        )

    def __eq__(self, other: object) -> bool:
        """Two off-targets are equal when they map to the same genomic locus.

        Identity is defined by chromosome, position, strand, and the aligned
        spacer sequence (which fully determines the target window).  Score
        fields are intentionally excluded so that the same hit scored by
        different methods compares equal.

        Parameters
        ----------
        other:
            Object to compare against.

        Returns
        -------
        bool
            ``True`` when *other* is an :class:`OffTarget` with the same locus.
        """
        if not isinstance(other, OffTarget):
            return NotImplemented
        return (
            self._chrom == other._chrom
            and self._pos == other._pos
            and self._strand == other._strand
            and self._spacer == other._spacer
        )

    def __hash__(self) -> int:
        """Hash based on the locus identity (mirrors :py:meth:`__eq__`).

        Returns
        -------
        int
            Hash value suitable for use in sets and dict keys.
        """
        return hash((self._chrom, self._pos, self._strand, self._spacer))

    # ==========================================================================
    # Properties – getters
    # ==========================================================================

    @property
    def chrom(self) -> str:
        """Chromosome name of the off-target site."""
        return self._chrom

    @property
    def pos(self) -> int:
        """1-based genomic start position."""
        return self._pos

    @property
    def strand(self) -> str:
        """Strand of the off-target site (``'+'`` or ``'-'``)."""
        return self._strand

    @property
    def grna(self) -> str:
        """Aligned guide RNA sequence (with PAM placeholder Ns)."""
        return self._grna

    @property
    def spacer(self) -> str:
        """Aligned genomic target (spacer) sequence including PAM bases."""
        return self._spacer

    @property
    def mm(self) -> int:
        """Number of substitution mismatches."""
        return self._mm

    @property
    def bulge_type(self) -> str:
        """Bulge classification (``'X'``, ``'DNA'``, ``'RNA'``, or ``'DNA/RNA'``)."""
        return self._bulge_type

    @property
    def bdna(self) -> int:
        """Number of DNA-bulge bases (gap in guide)."""
        return self._bdna

    @property
    def brna(self) -> int:
        """Number of RNA-bulge bases (gap in target)."""
        return self._brna

    @property
    def total_score(self) -> int:
        """Total edit distance: mismatches + DNA bulges + RNA bulges.

        Returns
        -------
        int
            ``mm + bdna + brna``.
        """
        return self._mm + self._bdna + self._brna

    @property
    def has_bulge(self) -> bool:
        """``True`` when the off-target involves at least one bulge position.

        Returns
        -------
        bool
            ``True`` if *bulge_type* is not ``'X'``.
        """
        return self._bulge_type != "X"

    @property
    def cfd_score(self) -> str:
        """CFD score string, or ``'NA'`` when not yet computed."""
        return self._cfd_score

    @cfd_score.setter
    def cfd_score(self, value: float) -> None:
        """Set the CFD score.

        Parameters
        ----------
        value:
            A numeric string (e.g. ``'0.85'``) or ``'NA'``.

        Raises
        ------
        TypeError
            If *value* is not a :class:`str`.
        """
        if not isinstance(value, float):
            raise TypeError(f"cfd_score must be a str, got {type(value).__name__!r}")
        self._cfd_score = f"{value:.2f}"

    @property
    def debug(self) -> bool:
        """Debug mode flag."""
        return self._debug

    # ==========================================================================
    # Serialisation
    # ==========================================================================

    def to_tsv_row(self, sep: str = "\t") -> str:
        """Serialise the off-target to a single delimited row.

        The column order matches :data:`TSV_HEADER` and is stable across all
        :class:`OffTarget` instances, making it safe to concatenate rows
        directly without a per-row header.

        Parameters
        ----------
        sep:
            Field separator (default: tab ``'\\t'``).

        Returns
        -------
        str
            A single line **without** a trailing newline.

        Examples
        --------
        >>> row = ot.to_tsv_row()
        >>> fields = row.split("\\t")
        >>> assert len(fields) == len(TSV_HEADER)
        """
        fields = [
            self._chrom,
            str(self._pos),
            self._strand,
            self._grna,
            self._spacer,
            str(self._mm),
            self._bulge_type,
            str(self._bdna),
            str(self._brna),
            self._cfd_score,
        ]
        return sep.join(fields)

    # ==========================================================================
    # Convenience helpers
    # ==========================================================================

    def cfd_score_as_float(self) -> Optional[float]:
        """Return the CFD score as a :class:`float`, or ``None`` if not set.

        Returns
        -------
        float or None
            Parsed CFD value, or ``None`` when :py:attr:`cfd_score` is
            ``'NA'`` or cannot be parsed as a float.
        """
        if self._cfd_score == _SCORE_NA:
            return None
        try:
            return float(self._cfd_score)
        except ValueError:
            return None

    def locus(self) -> str:
        """Return a compact human-readable locus string.

        Returns
        -------
        str
            E.g. ``'chr1:123456(+)'``.
        """
        return f"{self._chrom}:{self._pos}({self._strand})"
